### ISSUE TITLE
Mac crash

### ISSUE THREAD

**Issue by jerlinn** [huchenlei/ComfyUI-layerdiffuse#1]
Docode node always crash

💻 Mac M2

❌ `export PYTORCH_ENABLE_MPS_FALLBACK=1`
❌ `--force-fp16`
❌ just `python main.py`

![CleanShot 2024-03-03 at 14 44 43@2x](https://github.com/huchenlei/ComfyUI-layerdiffusion/assets/91647085/8e70a4ff-7f98-4c17-b109-1983f6a5e653)

**Comment by ynie** [comment]
Same running into this issue :(

**Comment by yiwangsimple** [comment]
M1 Same running into this issue :( +1

**Comment by huchenlei** [comment]
I do not have a Macbook with M-series chip. Can you help confirm that whether the issue exists for SD Forge's impl as well? https://github.com/layerdiffusion/sd-forge-layerdiffusion

**Comment by yiwangsimple** [comment]
![截屏2024-03-05 10 35 36](https://github.com/huchenlei/ComfyUI-layerdiffuse/assets/121092793/03f64daa-441c-4a54-8fcb-3991268bba9c)
My observation today is that it's the error that occurs when executing this node that causes Python to crash outright

**Comment by huchenlei** [comment]
Any Mac user tried https://github.com/layerdiffusion/sd-forge-layerdiffusion with any success?

**Comment by Tukeping** [comment]
@huchenlei Hi, Bro. Have you tried to fix this problem? I also have an MacOS with an M1 chip, and I have this problem.

**Comment by huchenlei** [comment]
> @huchenlei Hi, Bro. Have you tried to fix this problem? I also have an MacOS with an M1 chip, and I have this problem.

I would like to first confirm whether this issue is comfyui specific or for SD Forge as well.

**Comment by Tukeping** [comment]
> > @huchenlei Hi, Bro. Have you tried to fix this problem? I also have an MacOS with an M1 chip, and I have this problem.
> 
> I would like to first confirm whether this issue is comfyui specific or for SD Forge as well.

I tried running sd-forge-layerdiffuse and it reported the same error message.

------ Loggers ------------- Below ---------
Running on local URL:  http://127.0.0.1:7860

To create a public link, set `share=True` in `launch()`.
model_type EPS
UNet ADM Dimension 2816
Startup time: 14.5s (prepare environment: 0.4s, import torch: 5.2s, import gradio: 1.6s, setup paths: 2.1s, other imports: 3.0s, load scripts: 1.1s, create ui: 0.4s, gradio launch: 0.6s).
Using split attention in VAE
Working with z of shape (1, 4, 32, 32) = 4096 dimensions.
Using split attention in VAE
extra {'cond_stage_model.clip_g.transformer.text_model.embeddings.position_ids', 'cond_stage_model.clip_l.transformer.text_model.embeddings.position_ids', 'cond_stage_model.clip_l.logit_scale', 'cond_stage_model.clip_l.text_projection', 'cond_stage_model.clip_g.logit_scale'}
To load target model SDXLClipModel
Begin to load 1 model
Moving model(s) has taken 0.00 seconds
Model loaded in 7.7s (load weights from disk: 0.6s, forge load real models: 5.7s, calculate empty prompt: 1.3s).
[Layer Diffusion] LayerMethod.FG_ONLY_ATTN
To load target model SDXL
Begin to load 1 model
Moving model(s) has taken 3.52 seconds
100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 20/20 [00:11<00:00,  1.69it/s]
To load target model AutoencoderKL██████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 20/20 [00:10<00:00,  1.81it/s]
Begin to load 1 model
Moving model(s) has taken 0.10 seconds
To load target model UNet1024
Begin to load 1 model
Moving model(s) has taken 0.27 seconds
100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 8/8 [00:03<00:00,  2.16it/s]
/AppleInternal/Library/BuildRoots/0032d1ee-80fd-11ee-8227-6aecfccc70fe/Library/Caches/com.apple.xbs/Sources/MetalPerformanceShaders/MPSNDArray/Kernels/MPSNDArraySort.mm:287: failed assertion `(null)" Axis = 4. This class only supports axis = 0, 1, 2, 3
'
./webui.sh: line 292: 13941 Abort trap: 6           "${python_cmd}" -u "${LAUNCH_SCRIPT}" "$@"
/opt/homebrew/Caskroom/miniconda/base/lib/python3.10/multiprocessing/resource_tracker.py:224: UserWarning: resource_tracker: There appear to be 1 leaked semaphore objects to clean up at shutdown
  warnings.warn('resource_tracker: There appear to be %d '

**Comment by gabrie** [comment]
M2 Same running into this issue :( +1

**Comment by jerlinn** [comment]
Maybe in MPS frameworks, it can not sort the `batch_zise` dimension.

**Comment by tilseam** [comment]
M2 Pro+32G RAM get the same issue.

**Comment by BannyLon** [comment]
我的也是，mac M2 工作流一旦运行到LayerDiffusion解码(RGBA)，python就会自动中断运行，跳出错误提示

**Comment by yiwangsimple** [comment]
The mac doesn't have to get hung up on it. You can replace the node function with another process.

**Comment by BannyLon** [comment]
> The mac doesn't have to get hung up on it. You can replace the node function with another process.

How to replace node function with another process.

**Comment by hike2008** [comment]
same error, so how to resolve it

**Comment by BannyLon** [comment]
Looking forward to professional solutions, thank you very much!

**Comment by BannyLon** [comment]
The popularity of plugins is so high, how can this common bug be ignored and there are no experts to help solve it!

**Comment by dingshanliang** [comment]
same issue with mac m2.

**Comment by nigo81** [comment]
M3 pro max + 128g get the same issue.

**Comment by franksuni** [comment]
same issue with Mac M2 ultra. 
it does seem to be an MPS issue. 
the workaround I got is to use a custom node called "ImageSelector". you can apply this directly after vae decoder to select the layer you want. however, this will result in losing the matted FG (RGBA result with mask). use the [Generate BG + FG + Blended together] workflow as an example, apply 3 ImageSelectors after vae decode, and set selected_index each to 1, 2, 3 on the three selectors. what you get can be the blended FG+BG， separate FG (with a gray background, not transparent), and the BG it generated. 
I suppose there can be methods to get the mask again either by using a different selector that supports the mask; or use applying seg methods to create a new mask based on the FG's gray background. haven't tested anything yet. but you are welcome to try it by yourself.

**Comment by devgdovg** [comment]
The workaround below works fine on my m1max macbook.

In file `lib_layerdiffusion/models.py`, find 
```
median = torch.median(result, dim=0).values
```
and modify like this
```
if self.load_device == torch.device("mps"):
            '''
            In case that apple silicon devices would crash when calling torch.median() on tensors
            in gpu vram with dimensions higher than 4, we move it to cpu, call torch.median()
            and then move the result back to gpu.
            '''
            median = torch.median(result.cpu(), dim=0).values
            median = median.to(device=self.load_device, dtype=self.dtype)
else:
            median = torch.median(result, dim=0).values
```
Save and restart ComfyUI.

PS:
I tried to make a pull request, but it turns out I have no access rights or something wrong with my git config
```
ERROR: Permission to huchenlei/ComfyUI-layerdiffuse.git denied to devgdovg.
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
```
Anyone can help me with this? thanks.

**Comment by devgdovg** [comment]
to reproduce the crash in a simple scenario, try the code below on your m series macbook:
```
import torch
a = torch.randn(8, 1, 4, 512, 512)
mps_device = torch.device("mps")
b = a.to(mps_device)
tt = torch.median(b, dim=0) # crash here
```
but if you try a tensor with lower dimension, eg. `a = torch.randn(8, 1, 4, 512)`, there will be no crash

**Comment by forgetphp** [comment]
> The workaround below works fine on my m1max macbook.
> 
> In file `lib_layerdiffusion/models.py`, find
> 
> ```
> median = torch.median(result, dim=0).values
> ```
> 
> and modify like this
> 
> ```
> if self.load_device == torch.device("mps"):
>             '''
>             In case that apple silicon devices would crash when calling torch.median() on tensors
>             in gpu vram with dimensions higher than 4, we move it to cpu, call torch.median()
>             and then move the result back to gpu.
>             '''
>             median = torch.median(result.cpu(), dim=0).values
>             median = median.to(device=self.load_device, dtype=self.dtype)
> else:
>             median = torch.median(result, dim=0).values
> ```
> 
> Save and restart ComfyUI.
> 
> PS: I tried to make a pull request, but it turns out I have no access rights to this repo. I don't know if the repo's author have configured the repo settings.
> 
> ```
> ERROR: Permission to huchenlei/ComfyUI-layerdiffuse.git denied to devgdovg.
> fatal: Could not read from remote repository.
> 
> Please make sure you have the correct access rights
> ```
> 
> Anyone can help me with this? thanks.

我是 M1 用户。按照大佬的方法我成功的解决这个错误了。

<img width="1685" alt="image" src="https://github.com/huchenlei/ComfyUI-layerdiffuse/assets/13645582/3df473a4-9574-42d0-838f-4e20ba257a1d">

**Comment by tilseam** [comment]
> > The workaround below works fine on my m1max macbook.
> > In file `lib_layerdiffusion/models.py`, find
> > ```
> > median = torch.median(result, dim=0).values
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > and modify like this
> > ```
> > if self.load_device == torch.device("mps"):
> >             '''
> >             In case that apple silicon devices would crash when calling torch.median() on tensors
> >             in gpu vram with dimensions higher than 4, we move it to cpu, call torch.median()
> >             and then move the result back to gpu.
> >             '''
> >             median = torch.median(result.cpu(), dim=0).values
> >             median = median.to(device=self.load_device, dtype=self.dtype)
> > else:
> >             median = torch.median(result, dim=0).values
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Save and restart ComfyUI.
> > PS: I tried to make a pull request, but it turns out I have no access rights to this repo. I don't know if the repo's author have configured the repo settings.
> > ```
> > ERROR: Permission to huchenlei/ComfyUI-layerdiffuse.git denied to devgdovg.
> > fatal: Could not read from remote repository.
> > 
> > Please make sure you have the correct access rights
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Anyone can help me with this? thanks.
> 
> 我是 M1 用户。按照大佬的方法我成功的解决这个错误了。
> 
> <img alt="image" width="1685" src="https://private-user-images.githubusercontent.com/13645582/328008103-3df473a4-9574-42d0-838f-4e20ba257a1d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTUxNDU3MzEsIm5iZiI6MTcxNTE0NTQzMSwicGF0aCI6Ii8xMzY0NTU4Mi8zMjgwMDgxMDMtM2RmNDczYTQtOTU3NC00MmQwLTgzOGYtNGUyMGJhMjU3YTFkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTA4VDA1MTcxMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWE5OWNjMTFkYWJkYzJlNmE4ZDc0NTQ4MTJjY2Y4M2RjZWY3MWNlYjA4OTFmNTk2Yzc1MTlmMmE2ZTdhOTFhNDUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.GxAABHmetBVdoMKGJAe2Bpz6fYZfumj7BaCvW7TO53w">

小细节：不知道为什么用文本编辑器和xcode修改，插件会出错运行不了。用VSCode修改成功。

**Comment by tilseam** [comment]
> 
> 同是m1，能否上传你修改的文件，让我们覆盖原文件试试
[models.py.zip](https://github.com/huchenlei/ComfyUI-layerdiffuse/files/15292961/models.py.zip)
